oneflow/api/cpp/framework/graph.cpp

```diff
@@ -0,0 +1,256 @@
+
+namespace oneflow_api {
+
+namespace of = oneflow;
+
+namespace {
+
+class CompileScope {
+ public:
+  CompileScope(const of::JobConfigProto& job_config, const std::shared_ptr<of::NNGraph>& graph_) {
+    std::shared_ptr<of::Scope> scope = of::MakeInitialScope(job_config).GetOrThrow();
+    of::ThreadLocalScopeStackPush(scope).GetOrThrow();
+    of::JobBuildAndInferCtx_Open(job_config.job_name()).GetOrThrow();
+    of::cfg::JobConfigProto job_config_cfg(job_config);
+    of::CurJobBuildAndInferCtx_SetJobConf(job_config_cfg).GetOrThrow();
+  }
+
+  ~CompileScope() {
+    of::JobBuildAndInferCtx_Close().GetOrThrow();
+    of::ThreadLocalScopeStackPop().GetOrThrow();
+  }
+
+ private:
+  of::LazyMode::Guard lazy_mode_enabled_guard{true};
+};
+
+std::shared_ptr<of::one::TensorTuple> ConvertToTensorTuple(
+    const std::vector<std::shared_ptr<of::one::Tensor>>& tensors) {
+  auto tensor_tuple = std::make_shared<of::one::TensorTuple>();
+  for (const auto& tensor : tensors) { tensor_tuple->emplace_back(tensor); }
+  return tensor_tuple;
+}
+
+}  // namespace
+
+Graph::Graph(const std::string& model_path, const Device& device) : device_(device) {
+  // TODO(zzk0): model_path is a directory, need to concatenate filename
+  // we need a mlir model name.
+  // of::LoadJobFromIR(&job_, model_path).GetOrThrow();
+
+  std::ifstream input("/home/zhouzekai/oneflow/oneflow/ir/test/saved_model/job.pb");
+  std::stringstream buffer;
+  buffer << input.rdbuf();
+  job_.ParseFromString(buffer.str());
+  input.close();
+
+  graph_ = std::make_shared<of::NNGraph>(job_.job_conf().job_name());
+  of::Global<of::MultiClientSessionContext>::Get()->AddCGraph(graph_).GetOrThrow();
+}
+
+Graph::Graph(const std::string& model_path) : Graph(model_path, Device("cpu")) {}
+
+std::vector<Tensor> Graph::Forward(const std::vector<Tensor>& inputs) {
+  if (!is_compiled_) {
+    Compile(inputs);
+    is_compiled_ = true;
+  }
+  return Run(inputs);
+}
+
+void Graph::SetBatchSize(int batch_size) { batch_size_ = batch_size; }
+
+void Graph::Compile(const std::vector<Tensor>& inputs) {
+  BuildGraph(inputs);
+  LoadCheckpoint();
+  RegisterTensors();
+  graph_->CompileAndInitRuntime().GetOrThrow();
+}
+
+std::vector<Tensor> Graph::Run(const std::vector<Tensor>& inputs) {
+  auto input_tensor_tuple = std::make_shared<of::one::TensorTuple>();
+  for (const auto& tensor : inputs) { input_tensor_tuple->emplace_back(tensor.tensor_); }
+
+  of::RunLazyNNGraph(*input_tensor_tuple, *output_tensor_tuple_, *parameter_tensor_tuple_, graph_)
+      .GetOrThrow();
+  of::SoftSyncNNGraphBuffers(*output_tensor_tuple_, graph_).GetOrThrow();
+
+  std::vector<Tensor> outputs;
+  for (const auto& tensor : *output_tensor_tuple_) { outputs.emplace_back(Tensor(tensor)); }
+  return outputs;
+}
+
+void Graph::AddOp(oneflow::OperatorConf op_conf) {
+  std::shared_ptr<of::Scope> scope = oneflow::GetCurrentScope().GetPtrOrThrow();
+  op_conf.set_scope_symbol_id(scope->symbol_id().value_or(0));
```

Comment:
```suggestion
  {
    std::shared_ptr<of::Scope> scope = oneflow::GetCurrentScope().GetPtrOrThrow();
    op_conf.set_scope_symbol_id(scope->symbol_id().value_or(0));
  }
```
scope 这个变量只用了这一次，限制一下作用域好一些